bn_mp_radix_size_overestimate.c

```diff
@@ -0,0 +1,254 @@
+#elif ( (defined MP_28BIT) || (defined MP_31BIT) || (defined MP_32BIT)  || (defined MP_64BIT) )
+#define LTM_RADIX_SIZE_SCALE 61
+#ifdef MP_64BIT
+static const mp_digit logbases[65] = {
+#else
+static const mp_word logbases[65] = {
+#endif
+                      0ull,                    0ull,  2305843009213693952ull,
+    3654674702153732339ull,  4611686018427387904ull,  5354001665492895103ull,
+    5960517711367426291ull,  6473319721408181094ull,  6917529027641081856ull,
+    7309349404307464679ull,  7659844674706589055ull,  7976906213687625599ull,
+    8266360720581120243ull,  8532633055092329421ull,  8779162730621875046ull,
+    9008676367646627443ull,  9223372036854775808ull,  9425047617917838031ull,
+    9615192413521158631ull,  9795053960520401544ull,  9965687683920283007ull,
+   10127994423561913434ull, 10282749222901319551ull, 10430623713119086168ull,
+   10572203729794814195ull, 10708003330985790206ull, 10838476064306023373ull,
+   10964024106461197019ull, 11085005739835568998ull, 11201741516507896395ull,
+   11314519376860321395ull, 11423598928577852074ull, 11529215046068469760ull,
+   11631580915841357939ull, 11730890627131531983ull, 11827321386901076197ull,
+   11921035422734852583ull, 12012181624960263419ull, 12100896969734095496ull,
+   12187307757246061761ull, 12271530693133976959ull, 12353673836347420459ull,
+   12433837432775607386ull, 12512114650772492370ull, 12588592232115013503ull,
+   12663351069800359783ull, 12736466722332780120ull, 12808009872697200099ull,
+   12878046739008508147ull, 12946639442816362188ull, 13013846340199484158ull,
+   13079722320071570371ull, 13144319073519717325ull, 13207685337486676934ull,
+   13269867115674890971ull, 13330907879180520702ull, 13390848749049262950ull,
+   13449728662674133884ull, 13507584525721590347ull, 13564451351070528819ull,
+   13620362386074015347ull, 13675349229302017118ull, 13729441937791546026ull,
+   13782669125715645774ull, 13835058055282163712ull
+};
+#else
+#  ifdef _MSC_VER
+#    pragma message("mp_radix_size_overestimate: unknown or undefined MP_xBIT type")
+#  else
+#    warning "mp_radix_size_overestimate: unknown or undefined MP_xBIT type"
+#  endif
+#endif
+/* *INDENT-ON* */
+
+mp_err mp_radix_size_overestimate(const mp_int *a, const int base, int *size)
+{
+   mp_int bi_bit_count, bi_k;
+#if ( (defined MP_8BIT) || (defined MP_16BIT) )
+   mp_int bi_k_bis;
+#endif
+   int rem, bit_count;
+   mp_err e = MP_OKAY;
+
+   if ((base < 2) || (base > 64)) {
+      *size = 0;
+      return MP_VAL;
+   }
+   /* floor( log_2(a) ) + 1 */
+   bit_count = mp_count_bits(a);
+
+   if (bit_count == 0) {
+      *size = 2;
+      return MP_OKAY;
+   }
+
+   /* A small shortcut for powers of two. */
+   if (!(base&(base-1))) {
+      unsigned int x = (unsigned int)base;
+      int y;
+      for (y=0; (y < 7) && !(x & 1u); y++) {
+         x >>= 1;
+      }
+      *size = bit_count/y;
+      rem = bit_count - ((*size) * y);
```

Comment:
rem scope. Could you please keep the two functions in sync?

Reply:
speaking of "keep in sync" shouldn't `base` then be called `radix` or the other way around? :-)

Reply:
> speaking of "keep in sync" shouldn't base then be called radix or the other way around?

\*grmbl\*
;-)

> Could you please keep the two functions in sync?

I still would like to to deprecate `mp_radix_size` because it just calls `mp_ilogb` now and replace it with a note in the documentation (So it would be `MP_DEPRECATED(mp_ilogb)` and **not**  `MP_DEPRECATED(mp_radix_size_overestimate)`).

Reply:
No it is not just ilogb. It is also the sign and the 0 byte addition. I would like to keep it. What is the problem with it?

Reply:
> No it is not just ilogb. It is also the sign and the 0 byte addition. I would like to keep it. 
> What is the problem with it?

The main use case, determining the length of the *string* for memory allocation, has been made redundant by `mp_radix_size_overestimate`.

If you have a case where you need `mp_radix_size` and can not, by any means use `mp_ilogb` instead: tell me.

Otherwise it's around 2.5 Kb that can be spend better elsewhere.

Reply:
Determining string length is something which I imagine to be useful from time to time. Printing etc. 

Sure the user could also write ilogb+1+sign. But they would also have to check for 0? 

If someone cares about the kbs they will configure the lib to their needs, so nothing significant is being spent.

Reply:
> Determining string length is something which I imagine to be useful from time to time. Printing etc.

But that's *exactly* what `mp_radix_size` does and is meant for and described to do in the documentation.

> Sure the user could also write ilogb+1+sign.

Yes.

> But they would also have to check for 0?

Yes. And?

Does it happen that often that you need the exact value that it pays for a dedicated function?

Reply:
I say we keep it. @sjaeckel what do you think?

Reply:
> Determining string length is something which I imagine to be useful from time to time.

I ask myself why someone would want to use `mp_radix_size` after there's `mp_radix_size_overestimate`?